### Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

### Importing Data

In [2]:
# Folder Path

path = r'C:\Users\Windows 10\Documents\04-2023 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Windows 10\\Documents\\04-2023 Instacart Basket Analysis'

In [4]:
# Importing 'orders_products_merged_update.pkl'

df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_update.pkl'))

In [5]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly days        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

In [6]:
df_orders_products_merged.shape

(32404859, 18)

In [7]:
# Creating subset

df = df_orders_products_merged[:1000000]

In [8]:
df.shape

(1000000, 18)

In [9]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly days        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders  
5  Regularly busy  Regularly days        Average orders  
6  Regularly busy    Busiest days           Most orders  
7  Regularly busy    Busiest days           Most orders  
8  Regularly busy    Busiest days           Most orders  
9      Least busy    Slowest days        Average orders

### Grouping Data with Pandas

In [10]:
df.groupby('product_name')

In [ ]:
# To recap, you should always use the groupby() function as part of a series of steps, namely, the following:

# 1. Split the data into groups based on some criteria.
# 2. Apply a function to each group separately.
# 3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.

# So far, you’ve only completed the first step—splitting the data into groups (with the group being the “product_name” column)

### Aggregating Data with agg()

In [ ]:
#  If you were to calculate the mean of the “order_number” column grouped by the “department_id” column,
# you could quickly compare the average number of orders across each Instacart department.
# This will involve two of the steps in the three-step process introduced above:

# 1. Split the data into groups based on “department_id.”
# 2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

#### Single Aggregation

In [11]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [12]:
# Different way of finding mean of 'order_number'

df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

#### Multiple Aggregations

In [13]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

### Aggregating Data with transform()

In [ ]:
# Now, let’s map this task onto the three-step process introduced earlier:

# 1. Split the data into groups based on the “user_id” column.
# 2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# 3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

In [14]:
df_orders_products_merged['max_order'] = df_orders_products_merged.groupby('user_id')['order_number'].transform(np.max)
# df_orders_products_merged['max_order'] = df_orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

# Let’s run through each piece of the code in turn:
# First, a new column called “max_order” is created, which will be what stores the maximum order number for each user (step 3).
# Then, the df_orders_products_merged dataframe is grouped by the “user_id” column (step 1).
# And finally, the transform() function is applied on the “order_number” column with the np.max argument (step 2).

In [16]:
df_orders_products_merged.head(25)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   
20                     8.0         196                 14          1   
21                     4.0         196                  9          0   
22                     3.0         196                  1          1   
23                     NaN         196                  4          0   
24                     7.0         196                  5          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        7

### Deriving Columns with loc()

In [ ]:
# To create your flag, you’ll need some criteria. You can use the following:

# 1. If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# 2. If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# 3. If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [17]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [18]:
df_orders_products_merged.loc[(df_orders_products_merged['max_order'] <= 40) & (df_orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [19]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [20]:
df_orders_products_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [21]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly days        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [ ]:
# You learned already that to access a single column in a dataframe, you can index it like so: df['column']

In [26]:
# check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column

df_orders_products_merged[['user_id', 'max_order', 'loyalty_flag']].sample(30)

# The first set of brackets is the same as the brackets used in df['column'] above,
# while the second set of brackets is indicating that what you’re indexing is a list of multiple columns (and lists are always included within brackets).
# This is why there are two sets of brackets.

user_id  max_order      loyalty_flag
26336718   124336         35  Regular customer
14418039    39840         43    Loyal customer
27224138    86489          8      New customer
30315634    82347          3      New customer
424411     195887         25  Regular customer
14229437    34412         39  Regular customer
16104491   171470         66    Loyal customer
28076558    70802         34  Regular customer
19510421    55731         79    Loyal customer
16314835    74662         28  Regular customer
8863370     30658         12  Regular customer
11360235    68817         66    Loyal customer
30378956   135249         27  Regular customer
25876304   176981          7      New customer
5709740    181964         13  Regular customer
6905476     83605         34  Regular customer
30528089   162677         59    Loyal customer
25221661    59520         44    Loyal customer
20551180   180361          7      New customer
18108325    95224         34  Regular customer
26909498   165482         17  Regular customer
7164128     70037         24  Regular customer
11838936   175771         41    Loyal customer
6209100    119443         41    Loyal customer
3361181     60820          8      New customer
28474846    94042         10      New customer
24523973    76556         92    Loyal customer
23992020    21895         25  Regular customer
24988360   175719         19  Regular customer
19154405   106660          9      New customer

In [ ]:
# And that brings you to the end of the three-step process!
# 1. You started by grouping your data,
# 2. followed up by applying an aggregation function,
# 3. and finished everything off by using the resulting aggregation to create a flag column

### Task 4.8

### Step 1

#### Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [ ]:
# DONE

### Step 2

#### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [27]:
df_orders_products_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3

#### Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [28]:
# This is the subset one (I decided to run again instead of scrolling up)

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

#### Answer: The average result of order_number from the subset (1000000) is greater than the average result of entire dataframe (except the department_id: 17) 

### Step 4

#### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [ ]:
# DONE (Output [26])

### Step 5

#### Determine the difference between the spending habits of the three types of customers you identified.
#### Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [29]:
print('The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:')
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#### Observation:
#### 1. It is interesting to see that new customers have a higher average purchase price than the other 2 customer categories and even loyal customers have a lower average purchase than the others
#### 2. The maximum price here seems odd compared to the mean value itself. The value of 99.999 has skewed the distribution so further analysis for this value will be performed.


In [30]:
# Run descriptive statistics on the 'prices' column

print('Descriptive statistics of the prices column:')
df_orders_products_merged['prices'].describe()

Descriptive statistics of the prices column:


count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

#### Observation:
#### 1. The average (mean) price is 11.9 and the maximum price is 99.999, through this we can see that there is a huge difference in this data
#### 2. The standard deviation has been confirmed which has a std of 495.655 indicating that we have a lot of variation in this data (especially large values that are higher than the mean)

In [31]:
# Check for abnormal values in the 'prices' column

print('Check for abnormal values in the prices column:')
df_orders_products_merged['prices'].sort_values(ascending = False)

Check for abnormal values in the prices column:


29165979    99999.0
29165985    99999.0
29165519    99999.0
29165518    99999.0
29165517    99999.0
             ...   
7870751         1.0
7870750         1.0
7870749         1.0
7870748         1.0
30034999        1.0
Name: prices, Length: 32404859, dtype: float64

In [32]:
print('Check for abnormal values in the prices column:')

df_orders_products_merged['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

#### Observation: The 99.999 and 14.900 prices look weird (that's a lot higher than the others)

In [33]:
# Get the output of df_orders_products_merged when the value in the 'prices' column is 99999 or 14900

max_price = df_orders_products_merged.loc[df_orders_products_merged['prices'].isin([99999, 14900])]

In [35]:
# Sample output:

print('Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:')
max_price.sample(5)

Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:


order_id  user_id  order_number  orders_day_of_week  \
10031252   2164311    47416            51                   4   
10032467     68613   101319            30                   4   
10033567   1555940   151552             7                   1   
10033263   1515182   137858            17                   2   
10034612   1644129   198594             3                   6   

          order_hour_of_day  days_since_prior_order  product_id  \
10031252                 12                     3.0       21553   
10032467                 10                     3.0       21553   
10033567                 18                    15.0       21553   
10033263                 15                    17.0       21553   
10034612                 16                    30.0       21553   

          add_to_cart_order  reordered                      product_name  \
10031252                  8          1  Lowfat 2% Milkfat Cottage Cheese   
10032467                  4          1  Lowfat 2% Milkfat Cottage Cheese   
10033567                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10033263                 10          1  Lowfat 2% Milkfat Cottage Cheese   
10034612                 12          0  Lowfat 2% Milkfat Cottage Cheese   

          aisle_id  department_id   prices _merge     price_range_loc  \
10031252       108             16  14900.0   both  High-range product   
10032467       108             16  14900.0   both  High-range product   
10033567       108             16  14900.0   both  High-range product   
10033263       108             16  14900.0   both  High-range product   
10034612       108             16  14900.0   both  High-range product   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
10031252      Least busy    Slowest days           Most orders         64   
10032467      Least busy    Slowest days           Most orders         99   
10033567  Regularly busy    Busiest days        Average orders          9   
10033263  Regularly busy  Regularly days           Most orders         20   
10034612  Regularly busy  Regularly days           Most orders         10   

              loyalty_flag  
10031252    Loyal customer  
10032467    Loyal customer  
10033567      New customer  
10033263  Regular customer  
10034612      New customer

In [36]:
# Descriptive statistics on max_price

print('Descriptive statistics on max_price:')
max_price.describe()

# I need to scroll to the right to find the price, department_id and product_id

Descriptive statistics on max_price:


order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        5127.000000  5127.000000  5127.000000         5127.0   
mean            7.986347     0.613614   104.732592           16.0   
std             6.868889     0.486968     8.231346            0.0   
min             1.000000     0.000000    84.000000           16.0   
25%             3.000000     0.000000   108.000000           16.0   
50%             6.000000     1.000000   108.000000           16.0   
75%            11.000000     1.000000   108.000000           16.0   
max            72.000000     1.000000   108.000000           16.0   

             prices    max_order  
count   5127.000000  5127.000000  
mean   26485.547494    29.791691  
std    29186.636448    24.929207  
min    14900.000000     3.000000  
25%    14900.000000    12.000000  
50%    14900.000000    21.000000  
75%    14900.000000    41.000000  
max    99999.000000    99.000000

#### Observation: Incredibly high prices (14.900 and 99.999) both from the department_id 16 with product_id 21553 and 33664

In [37]:
# Find the product_name with product_id 21553 and 33664

df_orders_products_merged.loc[df_orders_products_merged['product_id'].isin([21553, 33664])]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29166209                  8                     4.0       33664   
29166210                  9                     2.0       33664   
29166211                 15                     8.0       33664   
29166212                 20                     4.0       33664   
29166213                 11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
29166209                  1          0             2 % Reduced Fat  Milk   
29166210                  1          1             2 % Reduced Fat  Milk   
29166211                 25          0             2 % Reduced Fat  Milk   
29166212                  8          0             2 % Reduced Fat  Milk   
29166213                 23          1             2 % Reduced Fat  Milk   

          aisle_id  department_id   prices _merge     price_range_loc  \
10030345       108             16  14900.0   both  High-range product   
10030346       108             16  14900.0   both  High-range product   
10030347       108             16  14900.0   both  High-range product   
10030348       108             16  14900.0   both  High-range product   
10030349       108             16  14900.0   both  High-range product   
...            ...            ...      ...    ...                 ...   
29166209        84             16  99999.0   both  High-range product   
29166210        84             16  99999.0   both  High-range product   
29166211        84             16  99999.0   both  High-range product   
29166212        84             16  99999.0   both  High-range product   
29166213        84             16  99999.0   both  High-range product   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
10030345  Regularly busy  Regularly days           Most orders         40   
10030346  Regularly busy  Regularly days           Most orders         40   
10030347  Regularly busy  Regularly days        Average orders          4   
10030348     Busiest day    Busiest days        Average orders          4   
10030349  Regularly busy    Busiest days        Average orders         16   
...                  ...             ...                   ...        ...   
29166209     Busiest day    Busiest days        Average orders         39   
29166210     Busie

#### Observation: 
#### 1. The price of 14.900 comes from Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
#### 2. The price of 99.999 comes from 2 % Reduced Fat Milk (with product_id 33664)

In [39]:
# Create a filter 'dairy' for records with the department_id value of 16, 
# and then 'dairy_without_weird' to exclude records (without very high prices)
# using ~ to exclude the data

dairy = df_orders_products_merged.loc[df_orders_products_merged['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id  order_number  orders_day_of_week  \
14349963    507918   206053            20                   1   
28362698   1731165    73618            12                   1   
11052468    493942   160332             6                   1   
17288291   1625716       59            10                   4   
3848450     161551    66104            12                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
14349963                 16                    22.0       33000   
28362698                 14                     7.0       22416   
11052468                  6                    15.0       27845   
17288291                  8                    11.0        4472   
3848450                  14                     7.0       22035   

          add_to_cart_order  reordered  \
14349963                 15          1   
28362698                 34          1   
11052468                  4          1   
17288291                  3          0   
3848450                  13          1   

                                               product_name  aisle_id  \
14349963                                  Pure Irish Butter        36   
28362698  O'Soy Fruit on the Bottom Strawberry Organic S...       120   
11052468                                 Organic Whole Milk        84   
17288291                   Toasted Coconut Almondmilk Blend        91   
3848450                         Organic Whole String Cheese        21   

          department_id  prices _merge    price_range_loc     busiest_day  \
14349963             16    12.2   both  Mid-range product  Regularly busy   
28362698             16    10.4   both  Mid-range product  Regularly busy   
11052468             16    12.8   both  Mid-range product  Regularly busy   
17288291             16     2.3   both  Low-range product      Least busy   
3848450              16     9.0   both  Mid-range product  Regularly busy   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
14349963    Busiest days           Most orders         60    Loyal customer  
28362698    Busiest days           Most orders         13  Regular customer  
11052468    Busiest days         Fewest orders          9      New customer  
17288291    Slowest days        Average orders         10      New customer  
3848450   Regularly days           Most orders         33  Regular customer

In [40]:
# Check to see if the filter is working properly

print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()

Maximum price in the dairy_without_weird filter:


15.0

#### Observation: After excluding unusual values, the max price becomes more reasonable

In [41]:
# First: Observation on 2 % Reduced Fat Milk (with product_id 33664)
# Create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird

reduced_fat_milk = dairy_without_weird.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')

In [42]:
print('Sample output from reduced_fat_milk:')
reduced_fat_milk.sample(10)

Sample output from reduced_fat_milk:


order_id  user_id  order_number  orders_day_of_week  \
14570451    201686   201035            32                   4   
20821129   2795116    59333             3                   3   
12944501   1221466   155104            21                   1   
20792105   2675633   144956            62                   5   
16010216   1209490    62824            22                   6   
11709810   2204707   200082             4                   4   
16007063   3406533    48137            20                   4   
12938367   3294519    22913            58                   3   
16009210   2786543    58812             8                   1   
16347743   1869037   168138             3                   6   

          order_hour_of_day  days_since_prior_order  product_id  \
14570451                 16                    16.0       23909   
20821129                 10                     5.0        2962   
12944501                 16                     9.0       35914   
20792105                 18                     5.0       19820   
16010216                  5                     8.0        5785   
11709810                  8                    22.0       32478   
16007063                 15                     9.0        5785   
12938367                 12                     3.0       35914   
16009210                 11                     8.0        5785   
16347743                 13                     5.0       40174   

          add_to_cart_order  reordered                     product_name  \
14570451                  7          1              2% Reduced Fat Milk   
20821129                  7          1    Milk, Reduced Fat, 2% Milkfat   
12944501                  6          1  Dha Omega 3 Reduced Fat 2% Milk   
20792105                  4          1    Grassmilk 2% Reduced Fat Milk   
16010216                  6          1      Organic Reduced Fat 2% Milk   
11709810                  2          1              Reduced Fat 2% Milk   
16007063                  2          1      Organic Reduced Fat 2% Milk   
12938367                  2          1  Dha Omega 3 Reduced Fat 2% Milk   
16009210                  1          1      Organic Reduced Fat 2% Milk   
16347743                  8          1      2% Reduced Fat Organic Milk   

          aisle_id  department_id  prices _merge    price_range_loc  \
14570451        84             16     9.2   both  Mid-range product   
20821129        84             16     2.7   both  Low-range product   
12944501        84             16     3.3   both  Low-range product   
20792105        84             16    12.6   both  Mid-range product   
16010216        84             16     7.9   both  Mid-range product   
11709810        84             16    13.9   both  Mid-range product   
16007063        84             16     7.9   both  Mid-range product   
12938367        84             16     3.3   both  Low-range product   
16009210        84             16     7.9   both  Mid-range product   
16347743        84             16    11.4   both  Mid-range product   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
14570451      Least busy    Slowest days           Most orders         32   
20821129  Regularly busy    Slowest days           Most orders         99   
12944501  Regularly busy    Busiest days           Most orders         43   
20792105  Regularly busy  Regularly days        Average orders         72   
16010216  Regularly busy  Regularly days         Fewest orders         58   
11709810      Least busy    Slowest days        Average orders          9   
16007063      Least busy    Slowest days           Most orders         37   
12938367  Regularly busy    Slowest days           Most orders         83   
16009210  Regularly busy    Busiest days           Most orders         35   
16347743  Regularly busy  Regularly days           Most orders         59   

              loyalty_flag  
14570451  Regular customer  
20821129    Loyal customer  
12944501    L

In [43]:
# Quick statistics in reduced_fat_milk

print('Quick statistics in reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics in reduced_fat_milk:


order_id        user_id   order_number  orders_day_of_week  \
count  2.022380e+05  202238.000000  202238.000000       202238.000000   
mean   1.711414e+06  103040.384151      18.528684            2.748702   
std    9.876774e+05   59620.791055      18.288088            2.022770   
min    5.000000e+00      11.000000       1.000000            0.000000   
25%    8.586890e+05   51496.000000       5.000000            1.000000   
50%    1.706996e+06  102925.000000      12.000000            3.000000   
75%    2.566178e+06  154743.000000      26.000000            5.000000   
max    3.421036e+06  206209.000000      99.000000            6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      202238.000000           190556.000000  202238.000000   
mean           13.181835               10.538445   18768.761009   
std             4.193794                8.380235   13345.840890   
min             0.000000                0.000000    1940.000000   
25%            10.000000                5.000000    5785.000000   
50%            13.000000                7.000000   19820.000000   
75%            16.000000               14.000000   32478.000000   
max            23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      202238.000000  202238.000000  202238.000000       202238.0   
mean            5.423051       0.782692      83.716626           16.0   
std             5.397505       0.412415       6.511429            0.0   
min             1.000000       0.000000      21.000000           16.0   
25%             2.000000       1.000000      84.000000           16.0   
50%             4.000000       1.000000      84.000000           16.0   
75%             7.000000       1.000000      84.000000           16.0   
max           137.000000       1.000000      91.000000           16.0   

              prices      max_order  
count  202238.000000  202238.000000  
mean        9.086900      35.834769  
std         3.064869      26.069071  
min         1.200000       3.000000  
25%         7.900000      14.000000  
50%         9.200000      30.000000  
75%        11.400000      51.000000  
max        14.700000      99.000000

#### Observation: Now the mean, min and max prices for products that contain 'Reduced Fat', '2' and 'Milk' in the name are: 9.08, 1.2, 14.7. Since this result makes more sense, the price of product ID 33664, named '2 % Reduced Fat Milk' will be reduced from 99.999 to 9.99

In [44]:
# Second: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird

cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [45]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)

Sample output from cottage_cheese:


order_id  user_id  order_number  orders_day_of_week  \
16191586    225931   103288            10                   5   
29898900   3032386    32699             2                   3   
30576369   1912399   184375             7                   4   
30576285   1543783   149033            13                   1   
16192113   2990407   140112             5                   3   
16192663   2586053   189617             7                   2   
16190757    278929    25225             4                   2   
16190750   1092547    25025            12                   3   
16190732   1915910    24108            12                   3   
29899327    436486   188721             2                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
16191586                 11                     7.0       27790   
29898900                 10                    15.0        6852   
30576369                 21                    30.0       11203   
30576285                 14                     3.0       11203   
16192113                 20                    16.0       27790   
16192663                 23                    30.0       27790   
16190757                 15                     9.0       27790   
16190750                  9                    29.0       27790   
16190732                  8                     6.0       27790   
29899327                  9                     7.0        6852   

          add_to_cart_order  reordered  \
16191586                  2          1   
29898900                  3          0   
30576369                  6          1   
30576285                  5          1   
16192113                  1          0   
16192663                  6          0   
16190757                  3          0   
16190750                  1          1   
16190732                  4          1   
29899327                  6          1   

                                               product_name  aisle_id  \
16191586        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29898900         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   
30576369  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...       108   
30576285  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...       108   
16192113        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16192663        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16190757        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16190750        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16190732        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
29899327         4% Milkfat Cottage Cheese Small Curd 22 OZ       108   

          department_id  prices _merge    price_range_loc     busiest_day  \
16191586             16     7.7   both  Mid-range product  Regularly busy   
29898900             16     8.6   both  Mid-range product  Regularly busy   
30576369             16     2.4   both  Low-range product      Least busy   
30576285             16     2.4   both  Low-range product  Regularly busy   
16192113             16     7.7   both  Mid-range product  Regularly busy   
16192663             16     7.7   both  Mid-range product  Regularly busy   
16190757             16     7.7   both  Mid-range product  Regularly busy   
16190750             16     7.7   both  Mid-range product  Regularly busy   
16190732             16     7.7   both  Mid-range product  Regularly busy   
29899327             16     8.6   both  Mid-range product  Regularly busy   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
16191586  Regularly days           Most orders         53    Loyal customer  
29898900    Slowest days           Most orders         13  Regular customer  
30576369    Slowest days        Average orders         23  Regular customer  
30576285    Busiest days           Most orders         56    Loyal customer  
16192113    Slowest days        Average orders     

In [46]:
# Quick statistics in cottage_cheese

print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number  orders_day_of_week  \
count  3.487000e+03    3487.000000   3487.000000         3487.000000   
mean   1.684106e+06  103778.688271     15.484084            2.849728   
std    9.785618e+05   58583.418992     16.808776            2.031540   
min    2.301000e+03      52.000000      1.000000            0.000000   
25%    8.438150e+05   55839.000000      4.000000            1.000000   
50%    1.673171e+06  107636.000000     10.000000            3.000000   
75%    2.535750e+06  151541.500000     20.000000            5.000000   
max    3.419042e+06  206072.000000     98.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        3487.000000             3218.000000   3487.000000   
mean           13.257241               12.898073  22348.791225   
std             4.297401                9.231128   8826.332190   
min             0.000000                0.000000    195.000000   
25%            10.000000                6.000000  11203.000000   
50%            13.000000               10.000000  27790.000000   
75%            16.000000               19.000000  27790.000000   
max            23.000000               30.000000  31030.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        3487.000000  3487.000000  3487.000000         3487.0   
mean            5.840551     0.697448   107.650703           16.0   
std             5.955836     0.459429     5.502322            0.0   
min             1.000000     0.000000    21.000000           16.0   
25%             2.000000     0.000000   108.000000           16.0   
50%             4.000000     1.000000   108.000000           16.0   
75%             8.000000     1.000000   108.000000           16.0   
max            57.000000     1.000000   108.000000           16.0   

            prices    max_order  
count  3487.000000  3487.000000  
mean      7.228219    30.531116  
std       1.757900    25.413902  
min       2.400000     3.000000  
25%       7.700000    11.000000  
50%       7.700000    22.000000  
75%       7.700000    43.000000  
max      11.800000    99.000000

#### Observation: Now the mean, min and max prices for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name are: 7.23, 2.4, 11.8. Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese' will be replaced from 14.900 and inputed by the mean value mentioned above

In [47]:
# Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

df_orders_products_merged['prices'] = df_orders_products_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [48]:
# Check the statistics of the new 'prices' column in df_orders_products_merged

print('The statistics of the new prices column in df_orders_products_merged:')
df_orders_products_merged['prices'].describe()

The statistics of the new prices column in df_orders_products_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

#### Observation: Now the mean, std, min and max results make more sense than Output [30]

### Step 5 (Updated: After this step --> replace({99999: 9.99, 14900: cottage_cheese['prices'].mean())


#### Determine the difference between the spending habits of the three types of customers you identified.
#### Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [52]:
# Basic statistics of df_orders_products_merged by 'loyalty_flag' with the updated 'prices' column

print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

#### Observation: Although the result remains the same (compared to Output [29]): new customer holds the highest purchase price followed by regular customer then loyal customer, now the mean value becomes lower and max value is more reasonable (instead of 99999, now 25)

### Step 6

#### Create a spending flag for each user based on the average price across all their orders using the following criteria:
    1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
    2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [53]:
# Creating “average_price” column using transform()

df_orders_products_merged['average_price'] = df_orders_products_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [54]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly days        Average orders        10   
1  Regularly busy    Slowest days        Average orders        10   
2  Regularly busy    Slowest days           Most orders        10   
3      Least busy    Slowest days        Average orders        10   
4      Least busy    Slowest days           Most orders        10   

   loyalty_flag average_price  
0  New customer      6.367797  
1  New customer      6.367797  
2  New customer      6.367797  
3  New customer      6.367797  
4  New customer      6.367797  

[5 rows x 21 columns]

In [55]:
# Rounding averages 

df_orders_products_merged['average_price'] = df_orders_products_merged['average_price'].round(2)

In [56]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly days        Average orders        10   
1  Regularly busy    Slowest days        Average orders        10   
2  Regularly busy    Slowest days           Most orders        10   
3      Least busy    Slowest days        Average orders        10   
4      Least busy    Slowest days           Most orders        10   

   loyalty_flag average_price  
0  New customer          6.37  
1  New customer          6.37  
2  New customer          6.37  
3  New customer          6.37  
4  New customer          6.37  

[5 rows x 21 columns]

In [57]:
# Creating a 'spending_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [58]:
# Creating a 'spending_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [59]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0         Soda  ...     9.0    both  Mid-range product  Regularly busy   
1         Soda  ...     9.0    both  Mid-range product  Regularly busy   
2         Soda  ...     9.0    both  Mid-range product  Regularly busy   
3         Soda  ...     9.0    both  Mid-range product      Least busy   
4         Soda  ...     9.0    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly days        Average orders        10  New customer   
1    Slowest days        Average orders        10  New customer   
2    Slowest days           Most orders        10  New customer   
3    Slowest days        Average orders        10  New customer   
4    Slowest days           Most orders        10  New customer   

   average_price spending_flag  
0           6.37   Low spender  
1           6.37   Low spender  
2           6.37   Low spender  
3           6.37   Low spender  
4           6.37   Low spender  

[5 rows x 22 columns]

In [62]:
# Check the frequency

df_orders_products_merged['spending_flag'].value_counts(dropna = False)

Low spender     32284501
High spender      120358
Name: spending_flag, dtype: int64

### Step 7

#### Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
    1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
    2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
    3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [63]:
# Creating “median_prior_orders” column using transform()

df_orders_products_merged['median_prior_orders'] = df_orders_products_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [64]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc     busiest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag average_price  \
0  Regularly days        Average orders        10  New customer          6.37   
1    Slowest days        Average orders        10  New customer          6.37   
2    Slowest days           Most orders        10  New customer          6.37   
3    Slowest days        Average orders        10  New customer          6.37   
4    Slowest days           Most orders        10  New customer          6.37   

   spending_flag median_prior_orders  
0    Low spender                20.5  
1    Low spender                20.5  
2    Low spender                20.5  
3    Low spender                20.5  
4    Low spender                20.5  

[5 rows x 23 columns]

In [65]:
# Creating a 'order_frequency_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [66]:
# Creating a 'order_frequency_flag' flag

df_orders_products_merged.loc[(df_orders_products_merged['median_prior_orders'] > 10) & (df_orders_products_merged['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [67]:
# Creating a 'order_frequency_flag' flag

df_orders_products_merged.loc[df_orders_products_merged['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [68]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day    busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly days   
1         Soda  ...  Mid-range product  Regularly busy    Slowest days   
2         Soda  ...  Mid-range product  Regularly busy    Slowest days   
3         Soda  ...  Mid-range product      Least busy    Slowest days   
4         Soda  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0        Average orders        10  New customer          6.37   Low spender   
1        Average orders        10  New customer          6.37   Low spender   
2           Most orders        10  New customer          6.37   Low spender   
3        Average orders        10  New customer          6.37   Low spender   
4           Most orders        10  New customer          6.37   Low spender   

   median_prior_orders   order_frequency_flag  
0                 20.5  Non-frequent customer  
1                 20.5  Non-frequent customer  
2                 20.5  Non-frequent customer  
3                 20.5  Non-frequent customer  
4                 20.5  Non-frequent customer  

[5 rows x 24 columns]

In [69]:
# Check the frequency

df_orders_products_merged['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [71]:
# Check the null values

print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
df_orders_products_merged[df_orders_products_merged['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

Records in df_orders_products_merged with null values in column order_frequency_flag:


user_id  days_since_prior_order  median_prior_orders  \
13645692   159838                     NaN                  NaN   
17251990   159838                     NaN                  NaN   
17622767   159838                     NaN                  NaN   
24138593   159838                     NaN                  NaN   
25880002   159838                     NaN                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

In [72]:
# Check user_id 159838

df_orders_products_merged[df_orders_products_merged['user_id'] == 159838]

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

            price_range_loc  busiest_day  busiest_days busiest_period_of_day  \
13645692  Mid-range product  Busiest day  Busiest days        Average orders   
17251990  Mid-range product  Busiest day  Busiest days        Average orders   
17622767  Low-range product  Busiest day  Busiest days        Average orders   
24138593  Mid-range product  Busiest day  Busiest days        Average orders   
25880002  Low-range product  Busiest day  Busiest days        Average orders   

         max_order  loyalty_flag average_price spending_flag  \
13645692         1  New customer          7.42   Low spender   
17251990         1  New customer          7.42   Low spender   
17622767         1  New customer          7.42   Low spender   
24138593         1  New customer          7.42   Low spender   
25880002         1  New customer          7.42   Low spender   

          median_prior_orders order_frequency_flag  
13645692                  NaN                  NaN  
17251990                  NaN                  NaN  
17622767                  NaN                  NaN  
24138593                  NaN                  NaN  
25880002                  NaN                  NaN  

[5 rows x 24 columns]

#### Observation: user_id 159838 seems to be a new customer. Therefore, I leave the value of NaN as it is

In [73]:
df_orders_products_merged.shape

(32404859, 24)

### Step 8

#### Ensure your notebook is clean and structured and that your code is well commented.

### Step 9

#### Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [74]:
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))